crates/accelerate/src/two_qubit_decompose.rs

```diff
@@ -0,0 +1,190 @@
+        let c = Complex::<f64>::from(self);
+        c.arg()
+    }
+}
+
+fn __weyl_coordinates(unitary: MatRef<c64>) -> (f64, f64, f64) {
+    let pi = PI;
+    let pi2 = PI / 2.0;
+    let pi4 = PI / 4.0;
+    let uscaled = Scale(C1 / unitary.determinant().powf(0.25)) * unitary;
+    let uup = transform_from_magic_basis(uscaled);
+    let mut darg: Vec<_> = (uup.transpose() * &uup)
+        .complex_eigenvalues()
+        .into_iter()
+        .map(|x: c64| -x.arg() / 2.0)
+        .collect();
+    darg[3] = -darg[0] - darg[1] - darg[2];
+    let mut cs: Vec<_> = (0..3)
+        .map(|i| utils::modulo((darg[i] + darg[3]) / 2.0, 2.0 * pi))
+        .collect();
+    let cstemp: Vec<f64> = cs
+        .iter()
+        .map(|x| utils::modulo(*x, pi2))
+        .map(|x| x.min(pi2 - x))
+        .collect();
+    let mut order = utils::argsort(&cstemp);
+    (order[0], order[1], order[2]) = (order[1], order[2], order[0]);
+    (cs[0], cs[1], cs[2]) = (cs[order[0]], cs[order[1]], cs[order[2]]);
+
+    // Flip into Weyl chamber
+    let pi32 = 3.0 * pi2;
+    if cs[0] > pi2 {
+        cs[0] -= pi32;
+    }
+    if cs[1] > pi2 {
+        cs[1] -= pi32;
+    }
+    let mut conjs = 0;
+    if cs[0] > pi4 {
+        cs[0] = pi2 - cs[0];
+        conjs += 1;
+    }
+    if cs[1] > pi4 {
+        cs[1] = pi2 - cs[1];
+        conjs += 1;
+    }
+    if cs[2] > pi2 {
+        cs[2] -= pi32;
+    }
+    if conjs == 1 {
+        cs[2] = pi2 - cs[2];
+    }
+    if cs[2] > pi4 {
+        cs[2] -= pi2;
+    }
+    (cs[1], cs[0], cs[2])
+}
+
+// For debugging. We can remove this later
+#[pyfunction]
+#[pyo3(text_signature = "(unitary, /")]
+pub fn _weyl_coordinates(unitary: PyReadonlyArray2<Complex<f64>>) -> (f64, f64, f64) {
+    let u = unitary.as_array().into_faer_complex();
+    __weyl_coordinates(u)
+}
+
+#[pyfunction]
+#[pyo3(text_signature = "(basis_b, basis_fidelity, unitary, /")]
+pub fn _num_basis_gates(
+    basis_b: f64,
+    basis_fidelity: f64,
+    unitary: PyReadonlyArray2<Complex<f64>>,
+) -> usize {
+    let u = unitary.as_array().into_faer_complex();
+    __num_basis_gates(basis_b, basis_fidelity, u)
+}
+
+fn __num_basis_gates(basis_b: f64, basis_fidelity: f64, unitary: MatRef<c64>) -> usize {
+    let (a, b, c) = __weyl_coordinates(unitary);
+    let pi4 = PI / 4.0;
+    let traces = [
+        c64::new(
+            4.0 * (a.cos() * b.cos() * c.cos()),
+            4.0 * (a.sin() * b.sin() * c.sin()),
+        ),
+        c64::new(
+            4.0 * (pi4 - a).cos() * (basis_b - b).cos() * c.cos(),
+            4.0 * (pi4 - a).sin() * (basis_b - b).sin() * c.sin(),
+        ),
+        c64::new(4.0 * c.cos(), 0.0),
+        c64::new(4.0, 0.0),
+    ];
+    let mut imax: usize = 0;
+    let mut max_fid = 0.0;
+    for (i, trace) in traces.into_iter().enumerate() {
+        let fid = trace_to_fid(trace) * basis_fidelity.powi(i as i32);
+        if fid > max_fid {
+            max_fid = fid;
+            imax = i
+        }
+    }
+    imax
```

Comment:
You could do this a bit more idiomatically by using `fold` (https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) with something like:

```suggestion
    traces.into_iter().enumerate().fold((0, 0.0), |(imax, max_fid), (i, trace)| {
        let fid = trace_to_fid(trace) * basis_fidelity.powi(i as i32);
        if fid > max_fid {
            (i, fid)
        } else {
            (imax, max_fid)
        }
    }).0
```

(untested so there might be a syntax or other compile error)

Reply:
This is idomaticized in fecdd90